[3.3.5] Core/Entities: Template Optimizer for Locales

Langerz82 opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**
What if we created the users default locale QueryData, and created the others as requested during run-time? Just an idea, I think it could work with some changes to my code I made.
That way caching would work and it would have fast-loading times still, and a smaller memory footprint unless of course u have many locales playing during run-time.
ref:
https://github.com/TrinityCore/TrinityCore/issues/22126  

**Target branch(es):** 3.3.5/master

- [x] 3.3.5

**Issues addressed:** Closes #  (insert issue tracker number)
https://github.com/TrinityCore/TrinityCore/issues/22126


**Tests performed:** (Does it build, tested in-game, etc.)
Tested in-game with a enGB client.

**Known issues and TODO list:** (add/remove lines as needed)
- Needs to be tested on non-US clients and make sure Locales are still working fine for Items, Quests, Points of interest, Creature Names, and GameObjects.

<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

Langerz82 committed:
Movement Sync - 19271 Fix

Langerz82 committed:
Revert "Movement Sync - 19271 Fix"

This reverts commit 9b8c4856fd283a6f77c6eb658753e255021424cf.

Langerz82 committed:
Optimize Locale Templates

Langerz82 commented:
It didnt add the new files, and I removed them. Gonna have to do some of them again.

Langerz82 committed:
Locale Template Optimization - Missing Files

Langerz82 commented:
I got two "styles" of code the Quest templates uses an old style and arrays, the other use vectors to store the locale data and GetLocaleString. Let me know which one you think is better and I can change.

If it does not work on other client Locales its bugged and will need further fixing (I cant test on non-GB as I cant find the client).

Langerz82 commented:
errors bahhhh lemme fix

Langerz82 committed:
Update QuestDef.cpp

Langerz82 committed:
Update cs_npc.cpp

Langerz82 committed:
Update QuestDef.cpp

Langerz82 committed:
Update CreatureTemplate.cpp

Langerz82 committed:
Update Creature.cpp

Langerz82 committed:
Update cs_npc.cpp

Langerz82 committed:
Update CreatureData.h

ccrs commented on src/server/game/Server/Packets/QuestPackets.cpp:
```diff
@@ -23,6 +23,11 @@ void WorldPackets::Quest::QueryQuestInfo::Read()
 
 WorldPacket const* WorldPackets::Quest::QueryQuestInfoResponse::Write()
+{
+    return Write(LOCALE_enUS);
```
?

illfated commented on src/server/game/Entities/Creature/CreatureTemplate.cpp:
```diff
@@ -0,0 +1,171 @@
+/*
+* Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
+* Copyright (C) 2005-2009 MaNGOS <http://getmangos.com/>
```
Lines 2 through 17 should have 1 space indentation for all the Copyright headers.
I presume the code is unchanged (copied from other files), for this MaNGOS line.

illfated commented:
Title suggestion: **[3.3.5] Core/Entities: Template Optimizer for Locales**

Langerz82 committed:
Update ItemTemplate.cpp

Langerz82 committed:
Update ObjectMgr.cpp

Langerz82 committed:
Update QuestDef.cpp

Langerz82 committed:
Update QuestDef.cpp

Langerz82 committed:
Update QuestDef.h

Langerz82 committed:
Update QuestDef.cpp

Langerz82 commented:
Cleaned up some of my changes a bit. Does anyone know why cs_npc.cpp is still generating that error with appveyor?

illfated commented:
Sorry, I don't know. One of the real C++ coders in the TC team might know.
I can only guess it is an `#include` problem, but I could easily be wrong.

illfated commented on src/server/game/Entities/Creature/CreatureTemplate.h:
```diff
@@ -0,0 +1,175 @@
+class WorldPacket;
+namespace WorldPackets {
+    namespace Query {
+        class QueryCreatureResponse;
```
Please follow TC C++ codestyle, curly brackets get their own separate lines in core files:
```cpp
namespace WorldPackets
{
    namespace Query
    {
        class QueryCreatureResponse;
    }
}
```

illfated commented on src/server/game/Entities/GameObject/GameObjectData.h:
```diff
@@ -22,558 +22,18 @@
-        }
+namespace WorldPackets {
+    namespace Query {
+        class QueryGameObjectResponse;
```
codestyle

illfated commented on src/server/game/Entities/GameObject/GameObjectTemplate.cpp:
```diff
@@ -0,0 +1,80 @@
+/*
+* Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
+*
```
codestyle

illfated commented on src/server/game/Entities/GameObject/GameObjectTemplate.h:
```diff
@@ -0,0 +1,586 @@
+/*
+* Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
+*
```
codestyle

Langerz82 committed:
Update QuestDef.h

illfated commented on src/server/game/Entities/Item/ItemTemplate.h:
```diff
@@ -21,10 +21,16 @@
+
+namespace WorldPackets {
+    namespace Query {
+        class QueryItemSingleResponse;
```
codestyle

Langerz82 replied:
I'm not worried about semantics atm, the core structure of the code is the main concern but thanks for the heads up. ^_^

illfated commented on src/server/game/Entities/Creature/CreatureTemplate.h:
```diff
@@ -0,0 +1,175 @@
+/*
+* Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
+* Copyright (C) 2005-2009 MaNGOS <http://getmangos.com/>
```
codestyle

illfated commented on src/server/game/Globals/ObjectMgr.cpp:
```diff
@@ -4260,21 +4258,22 @@ void ObjectMgr::LoadQuests()
         QuestLoaderFunction LoaderFunction;
     };
 
+    
```
redundant blank line with whitespace

illfated commented on src/server/game/Quests/QuestDef.h:
```diff
@@ -278,8 +269,8 @@ class TC_GAME_API Quest
-        std::string ObjectiveText[QUEST_OBJECTIVES_COUNT];
-
+        std::string ObjectiveText[TOTAL_LOCALES][QUEST_OBJECTIVES_COUNT];
+        
```
whitespace

illfated commented:
Sorry for not being able to comment on the coding, but cleaning up the visual oddities makes it more readable for real coders.

Langerz82 committed:
Config Template Locale Generate

Langerz82 committed:
CreatureMovement::ToString include fix

Seyden commented:
Uhm.. why did you move so much code between the files? what purpose had it?

illfated commented:
Seems to be a consolidating operation, to isolate closely related code into their own files where the code seems to fit more naturally.

Langerz82 commented:
@Seyden it made sense to make some of them in there own Template files along with its data structures that were tightly coupled to the class.
Has anyone tested it on a non-US locale and is it working ok is the main issue.... Make sure you have emptied the cache folder in your Wow install directory so it tries to retrieve new non-US locales.

jackpoz commented:
The idea of this PR sounds quite complicated to me, I would leave it simple with the choices "cache everything at start" or "don't cache anything because we don't have enough ram".

WildCoffee commented:
It looks like this PR does two things. One, it splits a couple of internal-to-the-file classes into separate files. They're really only used by the classes of the files they already reside in. I don't see where this adds anything because they're not going to be used elsewhere. Two, it makes it so the server can cache multiple locale strings. I think a better approach to this would be to add a server variable that sets the default cached locale instead of defaulting to enUS. Most people connecting to a server are going to be a certain language (that's what Blizzard defines). This just seems unnecessarily complex.

Langerz82 commented:
1. I have to fix up the ItemTemplates to not use arrays as it's a bit of a hack.

2. Maybe we just have it so it caches the main language only or not at all. That way it's much faster doing one iteration per data item instead of 9  (I think we support atm). I agree that most Servers will be targeting one locale only.

3. Generating them on the fly is pretty damn fast i believe, since we fetch all language data from the DB and store in RAM anyway.

4. This patch was mainly trying to avoid duplicate non-locale data from being stored eating up ram. Agreed it's native data types but considering the amount of records the DB has eats quite a chunk.

Langerz82 committed:
[3.3.5] Core/Entities: Template Optimizer for Locales - Quest Update

Langerz82 committed:
Revert "[3.3.5] Core/Entities: Template Optimizer for Locales - Quest Update"

This reverts commit 9a8094ef3b27d35ce64c89da6ff79c88f9a3803f.

Langerz82 committed:
Quest Template Update #2

Langerz82 committed:
Quest Packet Write Fix

lost-illusion commented:
I think CACHE_DATA_QUERIES was made for "GET ME ALREADY BUILDED PACKET RIGHT NOW", 
If we cache queries by request it same as CACHE_DATA_QUERIES = 0, yes, later, we have ready cache for most popular items and performace should be ok, but... why ? We can cache, or didn't cache. 
If you have reasons for this (low RAM + long uptime), you can make for self, i mean, it dont "production" solution.
But maybe you can add this as one more config option ?
@Langerz82

Shauren commented:
I still don't get the point of this PR - the config is meant to shift time spent building the packets from runtime to startup and its doing exactly that, there is nothing to fix. If you value fast startups for development then disable it

Langerz82 committed:
Optimize Locale Templates

Locale Template Optimization - Missing Files

Update QuestDef.cpp
Update cs_npc.cpp
Update QuestDef.cpp
Update CreatureTemplate.cpp
Update Creature.cpp
Update cs_npc.cpp
Update CreatureData.h
Update ItemTemplate.cpp
Update ObjectMgr.cpp
Update QuestDef.cpp
Update QuestDef.cpp
Update QuestDef.h
Update QuestDef.cpp
Update QuestDef.h
Config Template Locale Generate

CreatureMovement::ToString include fix

[3.3.5] Core/Entities: Template Optimizer for Locales - Quest Update

Revert "[3.3.5] Core/Entities: Template Optimizer for Locales - Quest Update"

This reverts commit 9a8094ef3b27d35ce64c89da6ff79c88f9a3803f.

Quest Template Update #2

Quest Packet Write Fix

Langerz82 committed:
Merge remote-tracking branch 'origin/3.3.5' into 3.3.5

Langerz82 committed:
Template Locale - Fix Errors

Langerz82 commented:
Now that I've nearly finished it I'm seeing it as being a little redundant too as you can either cache all languages or not at all. It's a nice feature just to cache only the main language you use but not really needed for the production server. Perhaps we keep the pull open and if people end up liking this feature then we re-consider it?

Langerz82 committed:
Template Locale - Fix Errors 2

Langerz82 committed:
Locale Template - Quest Fixes, Creature Template Fix

illfated commented:
1. Why are you changing line 3 in the file `contrib/conf_merge/README` from `authserver.conf.dist` to `bnetserver.conf.dist` for the `3.3.5` branch?
2. Why are you adding `CREATE DATABASE hotfixes` to the 3.3.5 branch in `sql/create/create_mysql.sql` ?
3. Why have you removed NewLine/CR/LF at the end of the following 3 files:
    `sql/old/3.3.5a/TDB44_to_TDB45_updates/world/2011_12_13_00_world_misc.sql`
    `sql/old/3.3.5a/TDB44_to_TDB45_updates/world/2011_12_16_00_world_creature_template.sql`
    `sql/old/3.3.5a/TDB44_to_TDB45_updates/world/2011_12_16_00_world_spell_bonus_data.sql`

Much of this looks unrelated to the PR, does it really belong here?

Langerz82 commented:
Yep it's a mistake while using git, I'll fix soon thanks for pointing it out.

lost-illusion commented:
You can add config option to config - which locales should be cached on start, one, two, or all, or nothing.
And for all non-cached locales we can use another config option - build every time, or cache for future.
And, if we cache all locales at start - why we should for every opcode search in map and etc. 
This is strange.
